docs/design-documents/design_template.md

```diff
@@ -0,0 +1,194 @@
+For example, if you are trying to define an interface for configuring the device, you may describe
+the API as below.
+
+    `Configure API should have following signature:
+    mbed_error_status_t configure(int speed, enum flow_control_t flow_control, int parity_bits);`
+    where:`
+        speed - is an integer value representing the target communication speed
+        flow_control - is the enum value representing the flow control to be used
+        parity_bits - is an integer value indicating the parity_bits to be used
+       
+    And the function should return mbed_error_status_t value indicating the result of the call as below:
+        MBED_SUCCESS if the call is successful.
+        MBED_ERROR_INVALID_ARGUMENT if input values are invalid.`
+        MBED_ERROR_CONFIG_UNSUPPORTED if the device doesnt support the requested configuration.`
+
+**Configuration Sequence Diagram**  
+
+Sequence Diagrams, Data flow diagrams etc. For example, the flow chart for configure device API above may
+be something like below.
+
+![Configure device operation](./diagram_examples/flow_chart_example.jpg)
+
+###	Detailed Design 2 - e.g, `Starting and stopping the device`
+
+**API Description**
+Detailed API Description like signature of interface, explanation of arguments, return codes etc
+For example, for start and stop functions exported by the driver.
+
+**Sequence Diagrams for `Starting and stopping the device`**
+Sequence Diagrams, Data flow diagrams etc. For example, the sequence digram for stopping the device may look
+something like below.
+
+![Stopping the device](./diagram_examples/sequence_diagram_example.jpg)
+
+###	Detailed Design 3 - e.g, `Reading from and Writing to the device`
+
+**API Description**
+Detailed API Description like signature of interface, explanation of arguments, return codes etc
+
+**Sequence Diagram for `Reading from and Writing to the device`**
+Sequence Diagrams, Data flow diagrams etc.
+
+###	Detailed Design N - e.g, `Resetting the device`
+
+**API Description**
+Detailed API Description like signature of interface, explanation of arguments, return codes etc
+
+**Sequence Diagram for `Resetting the device`**
+Sequence Diagrams, Data flow diagrams etc.
+
+# Usage Scenarios and Examples
+Show pseudo-code or flowcharts explaining the usage of the feature. For example you may want to include 
+some pseudo code to demonstrate how the read functionality is used with the new driver design you are
+proposing.
+
+### Scenario 1 - e.g, `Reading from the device`
+Mention the specific usage scenarios here. For example, the below examples shows how to read data from
+device using the new APIs.
+
+**Scenario 1 Example 1. - e.g, `Reading a byte from device`**
+Mention the specific usage here. For example, the below example shows pseudo-code for how to read a single byte data from
+the device using the new APIs.
+
+```C
+char in_byte = 0;
+int num_bytes_read = read(&in_byte, 1);
+if(num_bytes_read == 1) {
```

Comment:
`if ()` - space there (check also the rest of the examples) so they follow our code style